File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.py
n, k = map(int, input().split())
bag = [[] for i in range(n+1)]

for _ in range(1, n+1):
    bag[_] = list(map(int, input().split()))

dp = [[0 for i in range(n+1)] for i in range(k+1)]

for i in range(1, k+1):
    weight = 0
    for j in range(1, n+1):
        if dp[i][j-1] == 0:
            if weight + bag[j][0] <= i:
                dp[i][j] = bag[j][1]
                weight = bag[j][0]
        else:
            dp[i][j] = dp[i][j-1]
            if i-bag[j][0] >= 0:
                dp[i][j] = max(dp[i][j], dp[i-bag[j][0]][j-1] + bag[j][1])

print(dp[k][n])

File: 백준/Gold/12865. 평범한 배낭/평범한 배낭.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int number = Integer.parseInt(st.nextToken());
        int weight = Integer.parseInt(st.nextToken());

        int[][] arr = new int[number+1][2];
        for (int i = 1; i <= number; i++){
            st = new StringTokenizer(br.readLine());
            arr[i][0] = Integer.parseInt(st.nextToken());
            arr[i][1] = Integer.parseInt(st.nextToken());
        }

        int[][] bag = new int[number+1][weight+1];

        for (int j = 1; j <= weight; j++){
            for (int i = 1; i <= number; i++) {
                if (bag[i-1][j] == 0) {
                    if (arr[i][0] <= j) bag[i][j] = arr[i][1];
                } else {
                    bag[i][j] = bag[i-1][j];
                    if (arr[i][0] <= j) {
                        bag[i][j] = Math.max(bag[i-1][j], bag[i-1][j-arr[i][0]] + arr[i][1]);
                    }
                }
            }
        }

        System.out.println(bag[number][weight]);
    }
}